src/main/java/com/redhat/devtools/lsp4ij/features/semanticTokens/viewProvider/LSPSemanticTokensFileViewProviderHelper.java

```diff
@@ -34,6 +35,7 @@
 public class LSPSemanticTokensFileViewProviderHelper implements LSPSemanticTokensContainer {
 
     private final LSPSemanticTokensFileViewProvider fileViewProvider;
+    private final ThreadLocal<Integer> effectiveOffsetPtr = new InheritableThreadLocal<>();
```

Comment:
And this is the second inheritable thread local, and it's the one that's used for sticky lines as the user scrolls the editor viewport. Together, these allows us to have a full effective offset regardless of whether or not it corresponds to a real PSI element (whether for semantic tokens or a real PSI structure) or not, even for language servers that don't support semantic tokens, e.g. CSS.